comp/collector/collector/collectorimpl/collector.go

```diff
@@ -75,3 +104,255 @@ func newNoneCollector() provides {
+func (c *collectorImpl) stop(_ context.Context) error {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if c.scheduler != nil {
+		c.scheduler.Stop() //nolint:errcheck
+		c.scheduler = nil
+	}
+	if c.runner != nil {
+		c.runner.Stop()
+		c.runner = nil
+	}
+	c.state.Store(stopped)
+	return nil
+}
+
+// RunCheck sends a Check in the execution queue
+func (c *collectorImpl) RunCheck(inner check.Check) (checkid.ID, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	ch := middleware.NewCheckWrapper(inner, c.senderManager)
+
+	var emptyID checkid.ID
+
+	if c.state.Load() != started {
+		return emptyID, fmt.Errorf("the collector is not running")
+	}
+
+	if _, found := c.checks[ch.ID()]; found {
+		return emptyID, fmt.Errorf("a check with ID %s is already running", ch.ID())
+	}
+
+	err := c.scheduler.Enter(ch)
+	if err != nil {
+		return emptyID, fmt.Errorf("unable to schedule the check: %s", err)
+	}
+
+	// Track the total number of checks running in order to have an appropriate number of workers
+	c.checkInstances++
+	if ch.Interval() == 0 {
+		// Adding a temporary runner for long running check in case the
+		// number of runners is lower than the number of long running
+		// checks.
+		c.log.Infof("Adding an extra runner for the '%s' long running check", ch)
+		c.runner.AddWorker()
+	} else {
+		c.runner.UpdateNumWorkers(c.checkInstances)
+	}
+
+	c.checks[ch.ID()] = ch
+	c.notify(ch.ID(), collector.CheckRun)
+	return ch.ID(), nil
+}
+
+// StopCheck halts a check and remove the instance
+func (c *collectorImpl) StopCheck(id checkid.ID) error {
+	if !c.started() {
+		return fmt.Errorf("the collector is not running")
+	}
+
+	ch, found := c.get(id)
+	if !found {
+		return fmt.Errorf("cannot find a check with ID %s", id)
+	}
+
+	// unschedule the instance
+	err := c.scheduler.Cancel(id)
+	if err != nil {
+		return fmt.Errorf("an error occurred while canceling the check schedule: %s", err)
+	}
+
+	err = c.runner.StopCheck(id)
+	if err != nil {
+		// still attempt to cancel the check before returning the error
+		_ = c.cancelCheck(ch, c.cancelCheckTimeout)
+		return fmt.Errorf("an error occurred while stopping the check: %s", err)
+	}
+
+	err = c.cancelCheck(ch, c.cancelCheckTimeout)
+	if err != nil {
+		return fmt.Errorf("an error occurred while calling check.Cancel(): %s", err)
+	}
+
+	// remove the check from the stats map
+	expvars.RemoveCheckStats(id)
+
+	// vaporize the check
+	c.delete(id)
+
+	return nil
+}
+
+// cancelCheck calls Cancel on the passed check, with a timeout
+func (c *collectorImpl) cancelCheck(ch check.Check, timeout time.Duration) error {
+	done := make(chan struct{})
+
+	go func() {
+		ch.Cancel()
+		close(done)
+	}()
+
+	select {
+	case <-done:
+		return nil
+	case <-time.After(timeout):
+		return fmt.Errorf("timeout while calling check.Cancel() on check ID %s, timeout: %s", ch.ID(), timeout)
+	}
+}
+
+func (c *collectorImpl) get(id checkid.ID) (check.Check, bool) {
+	c.m.RLock()
+	defer c.m.RUnlock()
+
+	ch, found := c.checks[id]
```

Comment:
Interestingly, this doesn't seem to work:
```
comp/collector/collector/collectorimpl/collector.go:276:9: not enough return values
        have (*"github.com/DataDog/datadog-agent/comp/collector/collector/collectorimpl/internal/middleware".CheckWrapper)
        want (check.Check, bool)
```

I guess map subscripts don't behave the same as a function returning a tuple?

Reply:
My mistake, the context of multiple return values (what you originally had) makes the index operator return those 2 values. Otherwise, without that context, only 1 value is returned. Your original code is correct.